pkgs/development/tools/godot/3/default.nix

```diff
@@ -1,109 +1,165 @@
+      #
+      # This patch simply replaces the entire Scons environment (including the PATH) with that
+      # of the OS. This isn't as surgical as just fixing the PATH, but it seems to work, and
+      # seems to be the Nix community's current strategy when using Scons.
+      /SConstruct/dontClobberEnvironment.patch
+
+      # The version of detect.py in the godot source specifies both the compile flags and link flags
+      # for most of godot's dependencies, but for alsa, pulseaudio, and udev, it omits the link flags.
+      # It has comments by them explaining that the link flags are not needed because godot will dlopen
+      # these libraries. However, the dlopen call seems to fail for the Nix-based builds, and including
+      # the link flags seems to fix it. This patch does just that by passing `--libs` to pkg-config.
+      /detect.py/addLinkFlagsToAlsaAndUdev.patch
+    ] ++
+
+    # Adding link flags to pulseaudio is necessary to get the godot tool to work, but it seems to break
+    # the export templates. This doesn't seem to apply to udev or alsa for some reason. It is made
+    # optional so the export templates derivation can omit the link flags.
+    lib.optional self.shouldAddLinkFlagsToPulse /detect.py/addLinkFlagsToPulse.patch
+  );
 
   enableParallelBuilding = true;
+  buildPlatform = "x11";
+  shouldBuildTools = true;
+  buildTarget = "release_debug";
 
-  sconsFlags = [ "target=release_debug" "platform=x11" ];
-  preConfigure = ''
-    sconsFlags+=" ${
-      lib.concatStringsSep " "
-      (lib.mapAttrsToList (k: v: "${k}=${builtins.toJSON v}") options)
-    }"
-  '';
+  shouldUseLinkTimeOptimization = self.buildTarget == "release";
 
-  outputs = [ "out" "dev" "man" ];
+  sconsFlags = [
+    "platform=${self.buildPlatform}"
+    "tools=${lib.boolToString self.shouldBuildTools}"
+    "target=${self.buildTarget}"
+    "bits=${toString stdenv.hostPlatform.parsed.cpu.bits}"
+    "use_lto=${lib.boolToString self.shouldUseLinkTimeOptimization}"
+  ];
+
+  shouldWrapBinary = self.shouldBuildTools;
+  shouldInstallHeaders = self.shouldBuildTools;
+  shouldInstallShortcut = self.shouldBuildTools && self.buildPlatform != "server";
+
+  outputs = ["out" "man"] ++ lib.optional self.shouldBuildTools "dev";
+
+  builtGodotBinNamePattern = if self.buildPlatform == "server" then "godot_server.*" else "godot.*";
+
+  godotBinInstallPath = "bin";
+  installedGodotBinName = self.pname;
+  installedGodotShortcutFileName = "org.godotengine.Godot3.desktop";
+  installedGodotShortcutDisplayName = "Godot Engine 3";
 
   installPhase = ''
-    mkdir -p "$out/bin"
-    cp bin/godot.* $out/bin/godot3
+    runHook preInstall
+
+    echo "Installing godot binaries."
+    outbin="$out/$godotBinInstallPath"
+    mkdir -p "$outbin"
+    cp -R bin/. "$outbin"
+    mv "$outbin"/$builtGodotBinNamePattern "$outbin/$installedGodotBinName"
+
+    if [ $shouldWrapBinary ]; then
```

Comment:
I imagine these conditionals should be done like `if [ -n "$shouldWrapBinary" ]` to be clearer, but I'm not too concerned if they work as-is

Reply:
Apparently when attributes of a derivation are exported as environment variables, this is done using the builtin `toString` nix function.  When this function is applied to `true`, the result is `"1"`, and shows up as `1` within the environment variable.  When applied to `false`, the result is `""`, and shows up as empty string within the environment variable.

When you do `if [ 1 ]; then ...`, it is interpreted as true.  When you do `if [  ]; then ...`, it is interpreted as false.

So basically, you can just do `if [ $someNixBooleanAttribute ]; then ...` and it will work as expected.

Reply:
I just asked in discord about whether there is reason to prefer what you were suggesting over what I i did.  Someone said that while what I did will work, given the way nix bools are stringified, it is vulnerable to injection, and you shouldn't trust the environment not to be overridden.  The -n approach is resistant to that, so I'll go ahead and switch to that.

Reply:
From [posix manual](https://bs2manuals.ts.fujitsu.com/download/manual/20457):
>  -n not specified:
The meaning is the same as above, but the condition is easier to read if `-n` is specified.